fiftyone/core/metadata.py

```diff
@@ -382,28 +389,79 @@ def compute_metadata(
             default value ``fiftyone.config.show_progress_bars`` (None), or a
             progress callback function to invoke instead
     """
-    num_workers = fou.recommend_thread_pool_workers(num_workers)
+    logger.info("Computing metadata...")
 
     if sample_collection.media_type == fom.GROUP:
         sample_collection = sample_collection.select_group_slices(
             _allow_mixed=True
         )
 
-    if num_workers <= 1:
-        _compute_metadata(
-            sample_collection, overwrite=overwrite, progress=progress
-        )
-    else:
-        _compute_metadata_multi(
+    if not overwrite:
+        sample_collection = sample_collection.exists("metadata", False)
+
+    if (sample_count := len(sample_collection)) == 0:
+        logger.info("No samples to compute metadata for")
+        return
+
+    if num_workers == 1:
+        return _compute_metadata_single(
             sample_collection,
-            num_workers,
             overwrite=overwrite,
             progress=progress,
+            num_samples=sample_count,
+            skip_failures=skip_failures,
+            warn_failures=warn_failures,
         )
 
+    if not num_workers:
+        num_workers = fou.recommend_thread_pool_workers(num_workers)
+
+    mapper = focm.MapperFactory.create(
+        "process",
+        num_workers=num_workers,
+    )
+    metadata_iter = mapper.map_samples(
+        sample_collection,
+        _compute_metadata_map_fcn,
+        iter_fcn=_compute_metadata_iter_fcn,
+        skip_failures=skip_failures,
+        progress=progress,
+    )
+
+    collection = sample_collection._root_dataset._sample_collection
+
+    # Collect update operations
```

Comment:
The next few lines are basically a partial implementation of a `set_values()` function that accepts a generator as input. We should just build that method and then use it here!

If we kept following this pattern, we'd have dozens of implementations of this exact same code throughout the codebase, which would be 😢  Imagine we find a bug or improvement to this code; we want to make that fix in one place, not dozens (or worse, only fix it in one place).

Reply:
The goal was to not get into the weeds with set_values() so I went with this. Unlike other places we are computing something and then setting a known field name, so we don't need the flexibility of set_values() or the overhead of all the checks.

Reply:
Note that we had tested it initially with `sample_collection.set_values("metadata", values, key_field="id")` (without modifying set_values) and saw no improvement with just map_samples + values iterator

Reply:
There doesn't seem to be that much about metadata specifically. Perhaps this could be more generic and placed somewhere else. If this is an efficient pattern for setting values, it would be great for any subsequent `map_samples()` user to be able to use also.

Reply:
To not increase the scope of this PR and because there's a lot more that needs to be considered if handling other fields/user inputs, I've created a separate task to track that work: https://voxel51.atlassian.net/browse/FOEPD-1336